frontend/src/metabase/admin/settings/components/UploadSettings/UploadSettingsForm.tsx

```diff
@@ -0,0 +1,307 @@
+      "uploads-database-id": null,
+      "uploads-schema-name": null,
+      "uploads-table-prefix": null,
+    })
+      .then(() => {
+        setDbId(null);
+        setSchemaName(null);
+        setTablePrefix(null);
+        saveStatusRef?.current?.setSaved();
+      })
+      .catch(() => showError(disableErrorMessage));
+  };
+
+  const showPrefix = !!dbId;
+  const hasValidSettings = Boolean(dbId && (!showSchema || schemaName));
+  const settingsChanged =
+    dbId !== settings.uploads_database_id ||
+    schemaName !== settings.uploads_schema_name ||
+    tablePrefix !== settings.uploads_table_prefix;
+
+  const hasValidDatabases = databaseOptions.length > 0;
+  const isH2db = Boolean(
+    dbId && databases.find(db => db.id === dbId)?.engine === "h2",
+  );
+
+  return (
+    <PaddedForm aria-label={t`Upload Settings Form`}>
+      <Header />
+      {isH2db && <H2PersistenceWarning />}
+      <Group>
+        <Stack>
+          <SectionTitle>{t`Database to use for uploads`}</SectionTitle>
+          <Select
+            value={dbId ?? 0}
+            placeholder={t`Select a database`}
+            disabled={!hasValidDatabases}
+            options={databaseOptions}
+            onChange={(e: SelectChangeEvent<number>) => {
+              setDbId(e.target.value);
+              if (e.target.value) {
+                resetButtons();
+                dbHasSchema(databases, e.target.value)
+                  ? setTablePrefix(null)
+                  : setTablePrefix("upload_");
+                setSchemaName(null);
+              }
+            }}
+          />
+        </Stack>
+        {!!showSchema && (
+          <Schemas.ListLoader query={{ dbId, getAll: true }}>
+            {({ list: schemaList }: { list: Schema[] }) => (
+              <Stack>
+                <SectionTitle>{t`Schema`}</SectionTitle>
+                {schemaList?.length ? (
+                  <Select
+                    value={schemaName ?? ""}
+                    placeholder={t`Select a schema`}
+                    options={getSchemaOptions(schemaList)}
+                    onChange={(e: SelectChangeEvent<string>) => {
+                      resetButtons();
+                      setSchemaName(e.target.value);
+                    }}
+                  />
+                ) : (
+                  <EmptyState message={t`We couldn't find any schema.`} />
+                )}
+              </Stack>
+            )}
+          </Schemas.ListLoader>
+        )}
+        {!!showPrefix && (
+          <Stack>
+            <SectionTitle>{t`Upload Table Prefix (optional)`}</SectionTitle>
+            <Input
+              value={tablePrefix ?? ""}
+              placeholder={t`upload_`}
+              onChange={e => {
+                resetButtons();
+                setTablePrefix(e.target.value);
+              }}
+            />
+          </Stack>
+        )}
+      </Group>
+      <Group mt="lg">
+        {settings.uploads_enabled ? (
+          settingsChanged ? (
```

Comment:
i'm super not a fan of nested ternaries, something about the syntax breaks my brain. would appreciate if we could flatten this out, but i won't block you on it.